docs/old/reference/kernel_language.rst

```diff
@@ -0,0 +1,2050 @@
+
+```
+#pragma unroll 16 /* hint to compiler to unroll next loop by 16 */
+for (int i=0; i<16; i++) ...
+```
+
+```
+#pragma unroll 1  /* tell compiler to never unroll the loop */
+for (int i=0; i<16; i++) ...
+```
+
+
+```
+#pragma unroll /* hint to compiler to completely unroll next loop. */
+for (int i=0; i<16; i++) ...
+```
+
+
+## In-Line Assembly
+
+GCN ISA In-line assembly, is supported. For example:
+
+```
+asm volatile ("v_mac_f32_e32 %0, %2, %3" : "=v" (out[i]) : "0"(out[i]), "v" (a), "v" (in[i]));
+```
+
+We insert the GCN isa into the kernel using `asm()` Assembler statement.
+`volatile` keyword is used so that the optimizers must not change the number of volatile operations or change their order of execution relative to other volatile operations.
+`v_mac_f32_e32` is the GCN instruction, for more information please refer - [AMD GCN3 ISA architecture manual](http://gpuopen.com/compute-product/amd-gcn3-isa-architecture-manual/)
+Index for the respective operand in the ordered fashion is provided by `%` followed by position in the list of operands
+`"v"` is the constraint code (for target-specific AMDGPU) for 32-bit VGPR register, for more info please refer - [Supported Constraint Code List for AMDGPU](https://llvm.org/docs/LangRef.html#supported-constraint-code-list)
+Output Constraints are specified by an `"="` prefix as shown above ("=v"). This indicate that assemby will write to this operand, and the operand will then be made available as a return value of the asm expression. Input constraints do not have a prefix - just the constraint code. The constraint string of `"0"` says to use the assigned register for output as an input as well (it being the 0'th constraint).
+
+## C++ Support
+The following C++ features are not supported:
+- Run-time-type information (RTTI)
+- Try/catch
+- Virtual functions
+Virtual functions are not supported if objects containing virtual function tables are passed between GPU's of different offload arch's, e.g. between gfx906 and gfx1030. Otherwise virtual functions are supported.
+
+## Kernel Compilation
+hipcc now supports compiling C++/HIP kernels to binary code objects.
+The file format for binary is `.co` which means Code Object. The following command builds the code object using `hipcc`.
+
+`hipcc --genco --offload-arch=[TARGET GPU] [INPUT FILE] -o [OUTPUT FILE]`
+
+```
+[TARGET GPU] = GPU architecture
+[INPUT FILE] = Name of the file containing kernels
+[OUTPUT FILE] = Name of the generated code object file
+```
+
+Note: When using binary code objects is that the number of arguments to the kernel is different on HIP-Clang and NVCC path. Refer to the sample in samples/0_Intro/module_api for differences in the arguments to be passed to the kernel.
+
+## gfx-arch-specific-kernel
+Clang defined '__gfx*__' macros can be used to execute gfx arch specific codes inside the kernel. Refer to the sample 14_gpu_arch in samples/2_Cookbook.
```

Comment:
@MathiasMagnus I think we have newer examples available for this now. is that right?